dbms/src/Flash/Pipeline/PipelineTask.cpp

```diff
@@ -0,0 +1,135 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/Exception.h>
+#include <Common/MemoryTrackerSetter.h>
+#include <Flash/Pipeline/Event.h>
+#include <Flash/Pipeline/PipelineTask.h>
+
+namespace DB
+{
+PipelineTask::PipelineTask(
+    MemoryTrackerPtr mem_tracker_,
+    const EventPtr & event_,
+    OperatorExecutorPtr && op_executor_)
+    : Task(std::move(mem_tracker_))
+    , event(event_)
+    , op_executor(std::move(op_executor_))
+{
+    assert(event);
+    assert(op_executor);
+}
+
+PipelineTask::~PipelineTask()
+{
+    op_executor.reset();
+    assert(event);
+    event->finishTask();
+    event.reset();
+}
+
+#define HANDLE_CANCELLED                  \
+    if (unlikely(event->isCancelled()))   \
+    {                                     \
+        op_executor.reset();              \
+        return ExecTaskStatus::CANCELLED; \
+    }
+
+#define HANDLE_ERROR                                            \
+    catch (...)                                                 \
+    {                                                           \
+        op_executor.reset();                                    \
+        assert(event);                                          \
+        event->toError(getCurrentExceptionMessage(true, true)); \
+        return ExecTaskStatus::ERROR;                           \
+    }
+
+#define HANDLE_FINISHED                  \
+    case OperatorStatus::FINISHED:       \
+    {                                    \
+        op_executor.reset();             \
+        return ExecTaskStatus::FINISHED; \
+    }
+
+ExecTaskStatus PipelineTask::executeImpl()
+{
+    HANDLE_CANCELLED
```

Comment:
Compared to the original implementation, which checks for cancel in BlockInputStream, checking for canel in Task may be less responsive.

Reply:
has checked `exec_status.isCancelled()` in `OperatorPipeline`.